src/data/tree.lean

```diff
@@ -81,4 +84,69 @@ def map {β} (f : α → β) : tree α → tree β
 | nil := nil
 | (node a l r) := node (f a) (map l) (map r)
 
+/-- A non-nil `tree α`; useful when we want an arbitrary value other than `nil` -/
+abbreviation non_nil [inhabited α] : tree α := node default nil nil
+
+@[simp] lemma non_nil_ne [inhabited α] : non_nil ≠ (@nil α) := by trivial
```

Comment:
Is that really useful?

Reply:
It is used frequently in https://github.com/leanprover-community/mathlib/pull/16415/files. For example, the encoding of `bool` makes `tt` and `ff` to `nil` and `non_nil`. But since it's more pertinent to that PR than this one, I'll delete it.